include/slb/detail/invoke.hpp

```diff
@@ -0,0 +1,405 @@
+template <typename F, typename T1, typename... Tn>
+struct mem_fun_ptr_result<ptr_tag,
+                          F(T1, Tn...),
+                          decltype((
+                              void)(((*std::declval<T1>()).*std::declval<F&>())(
+                              std::declval<Tn>()...)))> {
+  using type = decltype(
+      ((*std::declval<T1>()).*std::declval<F&>())(std::declval<Tn>()...));
+};
+
+////////////////////////////////////////////////////////////////////////////////
+template <typename T>
+class mem_fun_ptr;
+
+template <typename T, typename C>
+class mem_fun_ptr<T C::*> {
+  T C::*_pm;
+
+private:
+  // `(t1.*pm)(t2, ..., tN)` when `pm` is a pointer to data member of a class
+  // `C` and `is_base_of_v<T, remove_reference_t<decltype(t1)>>` is `true`;
+  template <typename R, typename T1, typename... Tn>
+  static constexpr R _call(ref_tag, T C::*pm, T1&& t1, Tn&&... tn) noexcept(
+      noexcept((std::forward<T1>(t1).*pm)(std::forward<Tn>(tn)...))) {
+    return (std::forward<T1>(t1).*pm)(std::forward<Tn>(tn)...);
+  }
+
+  // `(t1.get().*pm)(t2, ..., tN)` when `pm` is a pointer to data member of a
+  // class `C` and `remove_cvref_t<decltype(t1)>` is a specialization of
+  // `reference_wrapper`;
+  template <typename R, typename T1, typename... Tn>
+  static constexpr R
+  _call(ref_wrapper_tag, T C::*pm, T1&& t1, Tn&&... tn) noexcept(
+      noexcept((t1.get().*pm)(std::forward<Tn>(tn)...))) {
+    return (t1.get().*pm)(std::forward<Tn>(tn)...);
+  }
+
+  // `((*t1).*pm)(t2, ..., tN)` when `pm` is a pointer to data member of a
+  // class `C` and `t1` does not satisfy the previous two items;
+  template <typename R, typename T1, typename... Tn>
+  static constexpr R _call(ptr_tag, T C::*pm, T1&& t1, Tn&&... tn) noexcept(
+      noexcept(((*std::forward<T1>(t1)).*pm)(std::forward<Tn>(tn)...))) {
+    return ((*std::forward<T1>(t1)).*pm)(std::forward<Tn>(tn)...);
+  }
+
+public:
+  constexpr mem_fun_ptr(T C::*pm) noexcept : _pm(pm) {}
+
+  template <typename T1,
+            typename... Tn,
+            typename PT1 = typename p0704<T, T1>::type,
+            typename Tag = typename dispatch_mem_ptr<C, PT1>::type,
+            typename R =
+                typename mem_fun_ptr_result<Tag, T C::*(PT1&&, Tn&&...)>::type,
+            typename = typename apply_clang_workaround<C, T, T1>::type>
+
+  constexpr R operator()(T1&& t1, Tn&&... tn) const noexcept(noexcept(
+      _call<R>(Tag{}, _pm, std::forward<PT1>(t1), std::forward<Tn>(tn)...))) {
+    return _call<R>(Tag{}, _pm, std::forward<PT1>(t1), std::forward<Tn>(tn)...);
+  }
+};
+
+} // namespace _invoke
+
+// We don't use `std::is_function<T>` here since some libstdc++ versions fail to
+// classify function types with noexcept specifiers.
+
+template <typename T>
+struct is_function {
+  static constexpr bool value =
+      !std::is_const<T const>::value && !std::is_reference<T>::value;
+};
+
+template <typename T, typename Enable = void>
+struct mem_fn_result {};
```

Comment:
Can we use `std::conditional` here?
```
template <typename T>
struct mem_fn_result {};

template <typename T, typename C>
struct mem_fn_result<T C::*> : std::conditional<is_function<T>::value,
                                                _invoke::mem_fun_ptr<T C::*>,
                                                _invoke::mem_obj_ptr<T C::*>> {
};
```

Reply:
We can, but as a base it would contribute to the entire class while it is only relevant for the nested `type` member. If instead it is restricted to the `type` alias definition, the compiler can immediately forget it rather than needing to keep track of it forever:

```
template <typename T>
struct mem_fn_result {};

template <typename T, typename C>
struct mem_fn_result<T C::*> {
  using type = typename std::conditional<is_function<T>::value,
      _invoke::mem_fun_ptr<T C::*>,
      _invoke::mem_obj_ptr<T C::*>>::type;
};
```